init.lua

```diff
@@ -146,19 +175,30 @@ function cudnn.toDescriptor(t)
    return descriptor
 end
 
+function cudnn.createDescriptors(count, descs_type, create_func, destroy_func)
+   local ds = ffi.new(descs_type, count)
+   for i = 0, count - 1 do
+      errcheck(create_func, ds + i)
+   end
+   local function destroyDescriptors(ds)
+      for i = 0, count - 1 do
+         errcheck(destroy_func, ds[i])
+      end
+   end
+   ffi.gc(ds, destroyDescriptors)
+   return ds
+end
 
 local sharedBuffer = {}
-for i=1,numDevices do
-    sharedBuffer[i] = {}
-end
 
 function cudnn.getSharedWorkspace()
     local device = cutorch.getDevice()
-    local stream = cutorch.getStream() -- starts from 0
-    if not sharedBuffer[device][stream] then
-       sharedBuffer[device][stream] = torch.CudaTensor(1)
+    if not sharedBuffer[device] then
```

Comment:
@borisfom, Thanks for your comments, I realize that what I describe is close to what your implementation is doing already. My suggestion was to have a utility function to (mostly) take guesswork out of figuring out maximum memory available for extraBuffer. It is also true that a particular convolution layer is not guaranteed to execute on the same stream. I'm not sure that reassigining buffer solves that. Suppose two convolutions start on the same stream and share same extraBuffer. Then one of them moves to another stream and that extraBuffer gets reassigned to that stream, another one stays in the same (initial) stream, comes online a bit afterwards, reassigns extraBuffer back to initial stream again - and boom! - two concurrently computed convolutions use the same extraBuffer. Or am I misunderstanding your suggestion?